File: Python/A_Project/A_Tic_Tac_Toe/GameStateHandler/Context/game_context.py
from GameStateHandler.ConcreteStates.x_turn_state import XTurnState


class GameContext:
    def __init__(self):
        self._current_state = XTurnState()

    def set_state(self, state):
        self._current_state = state

    def next(self, player, has_won: bool):
        self._current_state.next(self, player, has_won)

    def is_game_over(self) -> bool:
        return self._current_state.is_game_over()

    def get_current_state(self):
        return self._current_state

File: Ruby/A_Project/A_Tic_Tac_Toe/GameStateHandler/Context/game_context.rb
require_relative '../game_state'
require_relative '../ConcreteStates/x_turn_state'
require_relative '../ConcreteStates/o_turn_state'
require_relative '../ConcreteStates/x_won_state'
require_relative '../ConcreteStates/o_won_state'
require_relative '../ConcreteStates/draw_state'

module GameStateHandler
  module Context
    class GameContext
      attr_accessor :state

      def initialize
        # Start with X's turn
        @state = ConcreteStates::XTurnState.new
      end

      def next(player, has_won)
        @state.next(self, player, has_won)
      end

      def game_over?
        @state.game_over?
      end

      def current_state
        @state
      end
    end
  end
end
